File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
// class Solution {
// public:
//     int lengthOfLongestSubstring(string s) {
//         int low = 0, high=0, length = 0;
//         unordered_map<char,int>h_map;

//         while(high<s.size()){
            
//             h_map[s[high]]++;
            
//             while( h_map[s[high]]>1 && low<=high){
//              h_map[s[low]]--;
//              low++;
//             }
//             length = max(length,high-low+1); 
//             high++;
//         }

//         return length;
      
//     }
// };

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, high = 0, answer = 0;
        unordered_map<char, int>h_map;
        int temp = 0;
        while(high<s.size()){
            h_map[s[high]]++;
           

            if(h_map[s[high]]>1){
                answer = max(answer, temp);
                while(h_map[s[high]]>1){
                    h_map[s[left]]--;
                    temp--;
                    if(h_map[s[left]]==0) h_map.erase(s[left]);
                    left++;
                }
            }
            temp++;
            high++;
        }
        answer = max(answer, temp);
        return answer;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
         
        n = len(s)
        left = right = 0
        freq = {}
        longest = 0

        while (right<n):
            if s[right] not in freq:
                freq[s[right]] = 1
            else:
                freq[s[right]]+=1
            if freq[s[right]]>1:
                while (freq[s[right]]>1):
                    freq[s[left]]-=1
                    left+=1
            longest = max (longest, right-left+1)
            right += 1
        return longest

